### PR TITLE
Add filterMap macro

### PR DESCRIPTION
**Pull request by sebastiandedeyne**

### CODE REVIEW

**Comment by freekmurze** [comment]
Looks ok 👍 

Readme needs to be updated too.

File: tests/FilterMapTest.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+class FilterMapTest extends TestCase
+{
+    /** @test */
+    public function it_returns_a_mapped_collection_without_empty_values()
+    {
+        $result = Collection::make([1, 2, 3, 4, 5, 6])->filterMap(function ($number) {
+            $half = $number / 2;
+
+            return is_float($half)
```
Comment (jstoone) [inline tests/FilterMapTest.php]: Would it make sense to aim for the positive outcome:

```php
$quotient = $number / 2;

return is_integer($quotient)
  ? $quotient
  : null
```

In my example above I've changed `$half` to `$quotient`, as `$half` implies the quotient is not a whole number.

File: tests/FilterMapTest.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        $this->assertEquals([1, 2, 3], $result->values()->toArray());
+    }
+
+    /** @test */
+    public function it_returns_a_mapped_collection_with_a_custom_filter()
+    {
+        $result = Collection::make([1, 2, 3, 4, 5, 6])->filterMap(function ($number) {
+            return $number / 2;
+        }, function ($number) {
+            return ! is_float($number);
```
Comment (jstoone) [inline tests/FilterMapTest.php]: If we check whether the given `$number` is an integer there's no need for the negation.

```php
return is_integer($number);
```

**Comment by freekmurze** [comment]
>  flip the arguments so they come in the same order as the macro name

makes sense! 👍

**Comment by sebastiandedeyne** [comment]
My final thoughts on the naming:

`filterMap` sounds better — that's not an argument, but it's the reason why I'm about to make arguments 😅 

I'd say we drop the `$filter` argument, and simply see it as a variant on the `map` function. In that case, the order in the naming doesn't matter as much since there aren't parameters in the opposite order. If you need a `filter` callback, just chain. This is for the cases in which we want to map and dump extra values.

`flatMap` is also `map` then `flatten`, but it's called `flatMap` because the `map` is the main action. I'd give this method the same treatment.

**Comment by jstoone** [comment]
@sebastiandedeyne solid argument, and I'm solid. 👍

**Comment by freekmurze** [comment]
Agreeing with @sebastiandedeyne here. Make it so!

File: tests/FilterMapTest.php (hunk 1 above)
Comment (sebastiandedeyne) [reply tests/FilterMapTest.php]: Good call, updated!

**Comment by sebastiandedeyne** [comment]
Available in `3.1.0`!

**Pull request merged**